system/jlib/jfile.cpp

```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
+    std::unordered_map<int, Linked<CMonitoredItem>> monitoredFilesByFd;
+    std::unordered_map<std::string, Linked<CMonitoredItem>> monitoredFiles;
+    std::unordered_set<int> pendingRemovals;
+    CriticalSection crit;
+    FileWatchFunc callback;
+    CThreaded threaded;
+    bool stopped = true;
+
+    void handle_events(int fd)
+    {
+        byte buf[4096] __attribute__ ((aligned(__alignof__(struct inotify_event))));
+
+        while (true)
+        {
+            ssize_t len = read(fd, buf, sizeof(buf));
+            if (len == -1 && errno != EAGAIN)
+                throw makeErrnoException("CFileEventWatcher::handle_events - read");
+
+            if (len <= 0)
+                break;
+
+            const byte *ptr = buf;
+
+            do
+            {
+                const struct inotify_event *event = (const struct inotify_event *) ptr;
+                std::string filename;
+                if (event->len) // if watching directory event->name is populated with file within that dir.
+                    filename = event->name;
+                else
+                {
+                    // if watching a file, lookup filename from event handle
+                    CriticalBlock b(crit);
+                    auto iter = monitoredFilesByFd.find(event->wd);
+                    if (iter != monitoredFilesByFd.end())
+                    {
+                        filename = iter->second->filename;
+                        break;
+                    }
+                }
+                /* NB: in this impl. inotify_event flags are same as FileWatchEvents bit masks
+                 * If other implementations differ, they will need mapping to FileWatchEvents
+                 */
+                FileWatchEvents events = static_cast<FileWatchEvents>(event->mask);
+
+                callback(filename.c_str(), events);
+
+                ptr += sizeof(struct inotify_event) + event->len;
+            }
+            while (ptr < (buf + len));
+        }
+    }
+    void checkClose(int h)
+    {
+        if (-1 != h)
+            close(h);
+    }
+public:
+    CFileEventWatcher(FileWatchFunc _callback) : callback(_callback), threaded("CFileEventWatcher", this)
+    {
+        inotifyFd = inotify_init1(IN_NONBLOCK);
+        if (-1 == inotifyFd)
+            throw makeStringException(-1, "CFileEventWatcher - inotify_init1");
+        if (pipe(pipefd))
+            throw makeErrnoException(-1, "CFileEventWatcher - create pipe failed");
+        for (unsigned i=0; i<2; i++)
+        {
+            int flags = fcntl(pipefd[i], F_GETFL, 0);
+            if (-1 != flags)
+            {
+                flags |= O_NONBLOCK;
+                fcntl(pipefd[i], F_SETFL, flags);
+            }
+            flags = fcntl(pipefd[i], F_GETFD, 0);
+            if (-1 != flags)
+            {
+                flags |= FD_CLOEXEC;
+                fcntl(pipefd[i], F_SETFD, flags);
+            }
+        }
+
+        toPoll[0].fd = pipefd[0];
```

Comment:
Could do with some explanation about what the pipe is for.  I think it is to allow the watcher to be cancelled, but not at all clear.

Reply:
yes, agree, I'll comment to explain.